File: Regression/lin-reg.py
"""The purpose of this script is to perform simple linear regression on Walmart 
stock prices to determine the average growth over time."""

import numpy as np

days = []
highs = []

with open("test-data", "r") as data:
	for line in data:
		days.append(float(line.split(",")[0].strip()))
		highs.append(float(line.split(",")[1].strip()))

"""n = len(days)
x = np.array((days))
yTrain = np.array((highs))"""
n = 100
x = np.arange(100)
yTrain = np.arange(100)

slope = 0
yInter = 0

def linReg(x, y, alpha=0.0000001, epochs=10000):
	"""Return the slope, y intercept and mean squared error of a regression line 
	built on the given data.
	
	x, y = np arrays
	alpha = float
	epochs = int
	"""

	n = len(x)
	m, b = 0, 0

	for i in range(epochs):

		yTest = m * x + b
		error = yTest - y

		m -= alpha * 2 * np.sum(error * x) / n
		b -= alpha * 2 * np.sum(error) / n

	mSqrErr = np.sum(error ** 2) / n

	return m, b, mSqrErr


m, b, err = linReg(x, yTrain)
print(m, b, err)

"""learningRate = 0.0000001
epochs = 10000

print(yTrain.shape)

for i in range(epochs):

	yTest = slope * x + yInter

	error = yTest - yTrain
	squareError = error * error
	meanSquareError = np.sum(squareError) / n

	slope = slope - learningRate * 2 * np.sum(error * x) / n
	yInter = yInter - learningRate * 2 * np.sum(error) / n

	#if (i % 1000) == 0:
		#print(meanSquareError)

print("regression line: y =", slope, "x +", yInter)
print(meanSquareError)"""

File: Regression/linReg.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>

using namespace std;

class LinearRegression {
	vector<float> xTrain, yTrain;
	float slope, yInter;

	public:
		LinearRegression(vector<float> &xTrain, vector<float> &yTrain);
		void fitData(float alpha, int epochs);
		float getSlope();
		vector<float> predict(vector<float> &xTest);
		void addData(vector<float> &newX, vector<float> &newY);
		void resetData(vector<float> &newX, vector<float> &newY);
};

LinearRegression::LinearRegression(vector<float> &xTrain, vector<float> &yTrain) {
	this->xTrain = xTrain;
	this->yTrain = yTrain;
}

void LinearRegression::fitData(float alpha, int epochs) {
		/* Return the slope and y intercept of the regression line formed based on 
		   the given data. */

		slope = 0;
		yInter = 0;

		int elems = xTrain.size();
		vector<int> yTest(elems), error(elems);

		for (int i = 0; i < epochs; i++) {

				// Make predictions based on the current slope and y intercept.
				for (int j = 0; j < elems; j++) {
						yTest[j] = slope * xTrain[j] + yInter;
				}

				float errorSum = 0, errorSumX = 0;

				// Compare results to actual data and form the sum components of the 
				// gradient of the mean squared error function.
				for (int j = 0; j < elems; j++) {
						errorSum += yTest[j] - yTrain[j];
						errorSumX += (yTest[j] - yTrain[j]) * xTrain[j];
				}

				// Adjust parameters based on the gradient value and the learning
				// rate, alpha.
				slope -= 2 * alpha * errorSumX / elems;
				yInter -= 2 * alpha * errorSum / elems;
		}
}

float LinearRegression::getSlope() {
	return this->slope;
}

void LinearRegression::addData(vector<float> &newX, vector<float> &newY) {
	// Concatenate given vectors to xTrain and yTrain.

	this->xTrain.insert(xTrain.end(), newX.begin(), newX.end());
	this->yTrain.insert(yTrain.end(), newY.begin(), newY.end());
}

void LinearRegression::resetData(vector<float> &newX, vector<float> &newY) {
	// Replace xTrain and yTrain with given vectors.

	this->xTrain = newX;
	this->yTrain = newY;
}

float calcR2 (vector<float> &y, vector<float> &yTest) {
		/* Calculate the R-squared value for a given set of predictions.*/

		float ssRes = 0, ssTot = 0, R2, avg = 0;
		int n = y.size();

		for (int i = 0; i < n; i++) {
				avg += y[i];
		}

		avg /= n;

		// Compute the residual sum of squares and the total sum of squares.
		for (int i = 0; i < n; i++) {
				ssRes += pow(y[i] - yTest[i], 2);
				ssTot += pow(y[i] - avg, 2);
		}

		R2 = 1 - ssRes/ssTot;

		return R2;
}

float calcAvg (vector<float> &y) {
	int n = y.size();
	float avg = 0;

	for (int i = 0; i < n; i++) {
		avg += y[i];
	}

	avg /= n;
	
	return avg;
}
